# Warm up notebook for exploring RNA-seq data in transcripts per million
Data files:
* RNA-seq output from barrelseq pipeline
* mapping from sample id to experiment condition

In [1]:
import numpy as np
import pandas as pd

## Load tpm data and sample2condition mapping

In [2]:
# load TPM data
tpm_data_file = 'data/TPM_counts.tsv'
sample2condition_file = 'data/sample2condition.txt'

# load TPM data
df = pd.read_csv(tpm_data_file,sep='\t').fillna('')

    
# load mapping from sample to condition
with open(sample2condition_file,'r') as f:
    sample2condition = dict(x.strip().split() for x in f.readlines())


In [3]:
# View first 5 rows of the Dataframe
df.head()

locus_tag                                         product type  \
0  EQU24_RS00005  chromosomal replication initiator protein DnaA  CDS   
1  EQU24_RS00010                 DNA polymerase III subunit beta  CDS   
2  EQU24_RS00015             DNA replication/repair protein RecF  CDS   
3  EQU24_RS00020   DNA topoisomerase (ATP-hydrolyzing) subunit B  CDS   
4  EQU24_RS00025                            hypothetical protein  CDS   

  gene_symbol          locus  start_coord  end_coord  \
0        dnaA  NZ_CP035467.1            0       1317   
1              NZ_CP035467.1         1502       2603   
2        recF  NZ_CP035467.1         3060       4140   
3        gyrB  NZ_CP035467.1         4185       6600   
4              NZ_CP035467.1         6825       7062   

                                                note  \
0  Derived by automated computational analysis us...   
1  Derived by automated computational analysis us...   
2  Derived by automated computational analysis us...   
3  Derived by automated computational analysis us...   
4  Derived by automated computational analysis us...   

                                         translation  gene_len  ...  \
0  MSALWNNCLAKLENEISSSEFSTWIRPLQAIETDGQIKLLAPNRFV...      1318  ...   
1  MKYIINREQLLVPLQQIVSVIEKRQTMPILSNVLMVFRENTLVMTG...      1102  ...   
2  MSLQKLDIFNVRNIRQASLQPSPGLNLIYGANASGKSSVLEAIFIL...      1081  ...   
3  MSENIKQYDSTNIQVLKGLDAVRKRPGMYIGDTDDGTGLHHMVFEV...      2416  ...   
4  VKTTKYFLTTRMRPDREIIKDEWIQYVVRFPENEHIQFDGRIRRWA...       238  ...   

   5GB1_pA9_red_tpm  5GB1_pA9_yellow_tpm  5GB1C-5G-La-BR1_tpm  \
0         38.557373            38.810668            37.444214   
1         52.552767            52.461746            42.676553   
2         31.350991            34.914128            21.479309   
3         74.848501            80.850761            54.959319   
4         50.324948            49.349547            34.539657   

   5GB1C-5G-La-BR2_tpm  5GB1C-5G-N-BR1_tpm  5GB1C-5G-N-BR2_tpm  \
0            40.246006           40.100118           33.432274   
1            49.210083           46.798476           48.142385   
2            24.204682           22.171104           22.006566   
3            64.911376           59.653059           64.648318   
4            36.521074           37.789611           39.358066   

   5GB1C-JG15-La-BR1_tpm  5GB1C-JG15-La-BR2_tpm  5GB1C-JG15-N-BR1_tpm  \
0              39.880174              38.355431             30.247582   
1              45.465136              46.498139             37.152951   
2              22.658157              22.753325             19.407103   
3              69.119079              65.643179             57.590223   
4              38.992158              35.870964             41.462392   

   5GB1C-JG15-N-BR2_tpm  
0             41.248441  
1             52.902410  
2             29.834124  
3             68.306759  
4             40.227192  

[5 rows x 108 columns]

This data frame has 4213 rows: each row is a locus in the 5G genome for which RNA-seq data was captured.

In [4]:
df.shape

(4213, 108)

The data frame also has 108 columns! The first 10 columns have information about the locus, like its ID, it's type, the gene product it makes, its amino acid translation, its start and end coordinates, etc.

In [5]:
# columns with info about each locus
list(df.columns[:10])

['locus_tag',
 'product',
 'type',
 'gene_symbol',
 'locus',
 'start_coord',
 'end_coord',
 'note',
 'translation',
 'gene_len']

The remaining columns are sample names of different RNA-seq experiments run in the Lidstrom lab. Each column contains the Trascript Per Million (TPM) value for every gene in that particular experiment. 

In [6]:
# a few sample names from RNA-seq experiments
list(df.columns[11:21])

['5GB1_ferm_WT_QC_tpm',
 '5GB1_FM03_TR1_QC_tpm',
 '5GB1_FM03_TR2_QC_tpm',
 '5GB1_FM11_TR1_QC_tpm',
 '5GB1_FM11_TR2_QC_tpm',
 '5GB1_FM12_TR1_tpm',
 '5GB1_FM12_TR1_QC_tpm',
 '5GB1_FM12_TR2_tpm',
 '5GB1_FM12_TR2_QC_tpm',
 '5GB1_FM14_TR1_tpm']

In addition to the DataFrame, we also have a dictionary mapping each sample name to the "experimental condition" category to which it belongs. Experimental conditions help us group together which samples were run in similar biological conditions. Some of the samples are actual experimental replicates, but some are just experiments run by different people that had similar objectives. 

In [7]:
# python dictionary
sample2condition

{'5GB1_ferm_Ack_QC_tpm': 'lowO2_slow_growth',
 '5GB1_ferm_WT_QC_tpm': 'lowO2_slow_growth',
 '5GB1_FM03_TR1_QC_tpm': 'uMax',
 '5GB1_FM03_TR2_QC_tpm': 'uMax',
 '5GB1_FM11_TR1_QC_tpm': 'lowO2_fast_growth',
 '5GB1_FM11_TR2_QC_tpm': 'lowO2_fast_growth',
 '5GB1_FM12_TR1_tpm': 'lowCH4',
 '5GB1_FM12_TR1_QC_tpm': 'lowCH4',
 '5GB1_FM12_TR2_tpm': 'lowCH4',
 '5GB1_FM12_TR2_QC_tpm': 'lowCH4',
 '5GB1_FM14_TR1_tpm': 'lowCH4',
 '5GB1_FM14_TR1_QC_tpm': 'lowCH4',
 '5GB1_FM14_TR2_tpm': 'lowCH4',
 '5GB1_FM14_TR2_QC_tpm': 'lowCH4',
 '5GB1_FM18_TR1_QC_tpm': 'MeOH',
 '5GB1_FM18_TR2_tpm': 'MeOH',
 '5GB1_FM18_TR2_QC_tpm': 'MeOH',
 '5GB1_FM18_TR3_tpm': 'MeOH',
 '5GB1_FM18_TR3_QC_tpm': 'MeOH',
 '5GB1_FM19_TR1_tpm': 'lowO2_fast_growth',
 '5GB1_FM19_TR1_QC_tpm': 'lowO2_fast_growth',
 '5GB1_FM19_TR1_UW_tpm': 'lowO2_fast_growth',
 '5GB1_FM19_TR3_tpm': 'lowO2_fast_growth',
 '5GB1_FM19_TR3_QC_tpm': 'lowO2_fast_growth',
 '5GB1_FM20_TR1_QC_tpm': 'uMax',
 '5GB1_FM20_TR2_QC_tpm': 'uMax',
 '5GB1_FM20_TR3_tpm': 'uMax',
 '5G

# Data Exploration: Part 1
Start to explore the data in this pandas `DataFrame` object and answer the following questions:
1. How many loci are of each `type`?
1. What `product` is made by `locus_tag` "EQU24_RS19315"?
1. What is the longest locus?
1. How many loci exist between genome coordinates 100,000 and 200,000?
1. What's the highest expressed locus in the whole data frame?
    * What `type` of locus is it and what `product` does it make?
1. What's the highest expressed locus in of `type` "CDS"? Report `product` and `gene_symbol`.
    * Overall? 
    * For each sample?


In [8]:
# Your code and visualizations here

# Data Exploration: Part 2
Now let's think about adding experimental condition information to these samples. I find this easiest if we transpose the `DataFrame` so that the columns become genes (`locus_tag`s) and the rows are samples)

In [9]:
# list of all samples
samples = list(sample2condition.keys())

# transpose dataframe to make experiments the rows and genes the columns
df_T = df.set_index('locus_tag')[samples].T.reset_index().rename(columns={'index':'sample'})
# ^^ kinda gross pandas syntax... try to understand what each part does

In [10]:
df_T.head()

locus_tag                sample  EQU24_RS00005  EQU24_RS00010  EQU24_RS00015  \
0          5GB1_ferm_Ack_QC_tpm       2.933003       1.607784       1.415515   
1           5GB1_ferm_WT_QC_tpm       6.033848       3.895284       2.549771   
2          5GB1_FM03_TR1_QC_tpm      48.864921      51.315629      33.906257   
3          5GB1_FM03_TR2_QC_tpm      52.197450      54.947425      29.979783   
4          5GB1_FM11_TR1_QC_tpm      25.751902      37.216017      21.716802   

locus_tag  EQU24_RS00020  EQU24_RS00025  EQU24_RS00030  EQU24_RS00035  \
0               3.200081       1.522728       2.013385       8.890269   
1               4.301580       3.797084       1.694449       9.506517   
2              64.584281      74.801354      13.090237      50.320911   
3              72.717824      62.709260      24.874673      71.255094   
4              53.032824      46.348014      42.426259      58.772307   

locus_tag  EQU24_RS00040  EQU24_RS00045  ...  EQU24_RS22110  EQU24_RS22115  \
0               1.632474       0.790911  ...     170.826337       0.973237   
1               3.340407       2.160868  ...     259.349616       2.866741   
2              20.811187      21.910217  ...    6655.778183      56.948463   
3              19.773190      19.421485  ...    6480.237054      50.243190   
4               9.667881       8.212313  ...    6651.883413      25.635202   

locus_tag  EQU24_RS22120  EQU24_RS22125  EQU24_RS22130  EQU24_RS22135  \
0               2.344352       2.331287       4.867523       2.300157   
1               1.856930       0.951269       3.972334       3.217944   
2              52.002311      60.628466      23.015801     104.488835   
3              41.376658      66.581832      54.669612     103.864151   
4              29.059082      59.545627      34.189659      78.264069   

locus_tag  EQU24_RS22140  EQU24_RS22145  EQU24_RS22150  EQU24_RS22155  
0               5.261582       9.389775      21.657488      27.832080  
1              12.526353      21.072962      42.254357      41.530602  
2             379.686212     500.129230    1191.565348     796.964423  
3             401.742811     495.262554    1232.784967     724.291958  
4             208.094267     260.011454     666.589153     461.697526  

[5 rows x 4214 columns]

Now use pandas' apply function to add a column for `exp_condition`

In [11]:
# your code here (add exp_condition column)

Explore these questions:
* How many unique `exp_conditions` are there?
* How many samples belong to each `exp_condition`?

# Data Exploration Part 3

With this transposed `DataFrame` that incorporates the `exp_condition` column, we can more easily calculate the average TPM expression of each locus in each experimental condition. The new `DataFrame` will have the same number of columns (locus tags + sample name). But it should only have N rows (where N is the number of unique experimental conditions)

In [12]:
# your code here (create a new df containing the AVERAGE TPM experession of each locus_tag in each exp_condition)

With the `DataFrame` that now contains the average TPM expression across conditions, explore these questions:
* What are the top 5 loci in each `exp_condition`?
* In which `exp_condition` is "EQU24_RS19520" most highly expressed? Least highly expressed?
* For each `exp_condition`, what is the distribution of gene expression? (Try visualizing! Histogram, boxplot, violinplot, etc)


In [13]:
# Your code and visualizations here

## TODO

* correlation of samples within conditions
   * are there conditions we should exclude b/c poor quality? 
* coefficient of variation
* other QC?
* top n% in all conditions
* top n% in only 1 or 2 conditions